Make ocamlc preserve events after primitives even if they are at tail position.

jhjourdan opened the pull request:
This is a follow-up on the discussion in #9225: debug events occurring after a C call at tail position were dropped. The reason is that "after" debug events at tail position were systematically dropped to preserve the tail recursion optimization. This PR fixes this issue by checking that the after event does not appears around a primitive.

jhjourdan committed:
Make ocamlc preserve events after primitives even if they are at tail position.

gasche commented on testsuite/tests/statmemprof/intern.byte.reference:
```diff
@@ -1,15 +0,0 @@
-Called from file "intern.ml", line 32, characters 14-35
-Called from file "intern.ml", line 185, characters 2-25
-Called from file "intern.ml", line 191, characters 9-27
-OK !
```
I would expect this file to be deleted, but it is unclear from the github diff that it is.

jhjourdan replied:
It is indeed deleted.

gasche commented on bytecomp/bytegen.ml:
```diff
@@ -897,19 +897,29 @@ let rec comp_expr env exp sz cont =
-            | _                       -> Event_other
+            | Lprim((Pidentity | Popaque | Pdirapply | Prevapply), _, _) ->
+               true
+            | Lprim(_, _, _) -> false
```
I understand that it is a pain, but I would be more confident with an exhaustive listing of primitives rather than a fragile pattern-matching here, so that people adding new primitives have to consider this question. (Correspondingly, I would expect a comment explaining what this check is supposed to do, and I think it would be nicer to have it as a lam-parametrized toplevel function.)

jhjourdan replied:
Sorry, but this is *really* a pain: this algebraic type has more than 100 constructors!

I added a comment. As for new primitives: I would argue that it is very unlikely that such new primitives would require a change here. It is actually already arguable whether `f x` is a tail call in e.g., `Obj.magic (f x)` when this is in tail position.

gasche replied:
To finish reviewing and approve this PR, I feel that I should check that disabling tailcalls for all the primitives not listed above is correct. I guess that this involves looking at how each of them is compiled. Having a quick look at the list of primitives, I wonder about `Pignore, Praise, Psequand, Psequor, Pint_as_pointer`; could those not result in tailcalls in some situations?

jhjourdan replied:
`Praise` : this is compiled into the `Kraise` instruction, which never returns. The question of preserving the tail continuation is irrelevant.

`Pint_as_pointer`: this is compiled into a C call to the `caml_int_as_pointer` function

`Psequand`, `Psequor`: you're right. Fixed. (Although this seems to neverresult in a bug since an "event after" is never inserted around these primitives.)

gasche replied:
Thinking about this more, I think that it's one of these situations where we would regret not having written a complete pattern-matching, after the fact. I would like to list all primitives, and I think that it could be kept readable by moving it as a toplevel function (like `is_tailcall`) instead of being in the middle of the code. I'm happy to push a new commit to do this myself if you don't have the time.

jhjourdan replied:
Alright, let me do that.

gasche commented on bytecomp/bytegen.ml:
```diff
@@ -897,19 +897,29 @@ let rec comp_expr env exp sz cont =
-          comp_expr env lam sz cont1
+          if not preserve_tailcall && is_tailcall cont then
+            (* don't destroy tail call opt *)
+            comp_expr env lam sz cont
```
Sorry for the naive question, but I don't understand the `not preserve_tailcall` test here. My understanding from the `preserve_tailcall` code is that when it returns `true`, then we *should* preserve tail calls by *not* emitting debug information. But the test here suggests that you skip debug information when `not preserve_tailcall` holds, which is the opposite of what I would expect. (Again, an explanation comment could clarify.)

jhjourdan replied:
Indeed, you are right. This is completely bogus (and made the testsuite fail in several ways, btw). This is a result of a wrong last-minute refactoring. This is now fixed.

gasche commented:
One formulation that could be slightly clearer (easier for new primitive authors to follow correctly) is "is this primitive compiled into a C call?". If I understand correctly, the primitives for which we want to preserve debug events are those that turn into C calls, which never use the tailcall convention.

jhjourdan commented:
I did was you just asked : the pattern matching is now exhaustive. I don't find that particularly elegant, but I agree with the maintainability argument.

I also added another related change in the PR: I made the pattern matching of `Translprim.primitive_is_ccall` exhaustive, renamed that function, and included more primitives.

gasche reviewed (approved):
Thanks! The result is nice.

(The naming choice is slightly odd with `lambda_primitive_needs_event_after prim` on one hand and `preserve_tailcall_for_prim prim` on the other, but we're past the point of diminishing return and I'm happy to merge and let you move to something else.)
